File: src/UnEInternal/UnEInternal/Controllers/VacationController.cs
using System;
using System.Collections;
using System.Collections.Generic;
using Microsoft.Extensions.Configuration;
using Microsoft.AspNetCore.Mvc;
using Vacation.Model;
using Vacation.BLL;
using Vacation.IDAL;
using Vacation.BLL.Models.Account;
using Vacation.BLL.Models.Vacation;

// For more information on enabling Web API for empty projects, visit https://go.microsoft.com/fwlink/?LinkID=397860

namespace UnEInternal.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class VacationController : ControllerBase
    {
        private IConfiguration m_config = null;
        private ProcessManager m_processManager = null;

        public VacationController(IDataManager dataManager, IConfiguration config)
        {
            m_config = config;
            m_processManager = new ProcessManager(dataManager);
        }

        // POST api/<VacationController>
        [HttpPost]
        public IActionResult Post([FromBody] VacationData data)
        {
            if (data.RequestManager != null)
                return RequestManager(data.RequestManager);
            else if (data.RequestSpecialVacationManager != null)
                return RequestSpecialVacationManager(data.RequestSpecialVacationManager);
            else if (data.RequestHistory != null)
                return RequestHistory(data.RequestHistory);
            else if (data.RequestVacation != null)
                return RequestVacation(data.RequestVacation);
            else if (data.RequestSpecialVacation != null)
                return RequestSpecialVacation(data.RequestSpecialVacation);
            else if (data.RequestManagerData != null)
                return RequestManagerData(data.RequestManagerData);
            else if (data.ProcessRequest != null)
                return ProcessRequest(data.ProcessRequest);
            else if (data.RequestMemberHistory != null)
                return RequestMemberHistory(data.RequestMemberHistory);
            else if (data.RequestVacationList != null)
                return RequestVacationList(data.RequestVacationList);
            else if (data.RequestCancelVacations != null)
                return CancelVacations(data.RequestCancelVacations);
            else if (data.RequestHolidays != null)
                return RequestHolidays(data.RequestHolidays);

            return null;
        }

        private IActionResult RequestHolidays(RequestHolidays data)
        {
            ResponseHolidays response = data.Month == null ? HolidayManager.GetHolidays(data.Year, Startup.HolidayApiUrl, Startup.HolidayApiLicenseKey, Startup.CustomHolidays) : HolidayManager.GetHolidays(data.Year, (int)data.Month, Startup.HolidayApiUrl, Startup.HolidayApiLicenseKey, Startup.CustomHolidays);
            return Ok(response);
        }

        private IActionResult CancelVacations(RequestCancelVacations data)
        {
            ResponseCancelVacations result = Vacation.BLL.RequestManager.CancelVacations(m_processManager.GetDataManager(), data.RequestIDs, m_processManager.GetVacationManager());
            return Ok(result);
        }

        private IActionResult RequestVacationList(RequestVacationList data)
        {
            ResponseVacationList response = m_processManager.GetVacationManager().GetVacationList(data.UserID, data.Year);
            return Ok(response);
        }

        private IActionResult RequestHistory(_RequestHistory data)
        {
            Vacation.BLL.Models.Vacation.History history = m_processManager.GetVacationManager().GetVacationHistory(data.UserID, data.Year, data.Month, data.Day);
            return Ok(history);
        }

        private IActionResult RequestManager(_RequestManager data)
        {
            RequestManagerResult result = m_processManager.GetVacationManager().GetManager(data.UserID, data.GetDateList());
            return Ok(result);
        }

        private IActionResult RequestSpecialVacationManager(_RequestManager data)
        {
            RequestManagerResult result = m_processManager.GetVacationManager().GetSpecialVacationManager(data.UserID);
            return Ok(result);
        }

        private IActionResult RequestVacation(_RequestVacation data)
        {
            RequestVacationResult result = m_processManager.GetVacationManager().RequestVacation(data.UserID, data.GetDateList(), data.Description);
            return Ok(result);
        }

        private IActionResult RequestSpecialVacation(_RequestSpecialVacation data)
        {
            RequestSpecialVacationResult result = m_processManager.GetVacationManager().RequestSpecialVacation(data.RequestManagerID, data.UserIDs, data.Days, data.Reason);
            return Ok(result);
        }

        private IActionResult RequestManagerData(_RequestManagerData data)
        {
            ManagerRequestData result = m_processManager.GetVacationManager().RequestManagerData(data.ManagerUserID, data.Year);
            return Ok(result);
        }

        private IActionResult ProcessRequest(_ProcessRequest data)
        {
            ProcessRequestResult result = m_processManager.GetVacationManager().ProcessRequest(data.RequestID, data.IsPermit, data.ManagerUserID, data.ManagerDescription, data.IsNormal);
            return Ok(result);
        }

        private IActionResult RequestMemberHistory(_RequestMemberHistory data)
        {
            MemberVacationHistory result = m_processManager.GetVacationManager().GetMemberVacationHistory(data.ManagerUserID);
            return Ok(result);
        }
    }
}

File: src/UnEInternal/UnEInternal/ClientApp/src/Root/services/vacationController.js
import { JsonManager } from './jsonManager';

export class VacationController {
    static async requestManager(user, days) {
        try {
            const jsonData = JsonManager.makeRequestManager(user, days);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async requestSpecialVacationManager(user) {
        try {
            const jsonData = JsonManager.makeRequestSpecialVacationManager(user);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async requestHistory(user, year, month, day) {
        try {
            const jsonData = JsonManager.makeRequestHistory(user, year, month, day);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async requestVacation(user, days, description) {
        try {
            const jsonData = JsonManager.makeRequestVacation(user, days, description);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async requestSpecialVacation(manager, members, days, reason) {
        try {
            const userIDs = [];
            const memberCount = members.length;

            for (let i = 0; i < memberCount; i++) {
                const member = members[i];
                userIDs.push(member.userID);
            }

            const jsonData = JsonManager.makeRequestSpecialVacation(manager, userIDs, days, reason);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async requestManagerData(user, year) {
        try {
            const jsonData = JsonManager.makeRequestManagerData(user, year);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async processRequest(requestID, permit, managerUserID, managerDescription) {
        try {
            const jsonData = JsonManager.makeProcessRequest(requestID, permit, managerUserID, managerDescription, true);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async processSpecialVacationRequest(requestID, permit, managerUserID, managerDescription) {
        try {
            const jsonData = JsonManager.makeProcessRequest(requestID, permit, managerUserID, managerDescription, false);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async requestMemberHistory(managerUserID) {
        try {
            const jsonData = JsonManager.makeRequestMemberHistory(managerUserID);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                JsonManager.setMemberHistory(result);
                return result;
            }
        }
        catch (e) {
            console.log(e);
        }

        return null;
    }

    static async requestVacationList(userID, year) {
        try {
            const jsonData = JsonManager.makeRequestVacationList(userID, year);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return [result.success, result.message, result.vacations];
            }
        }
        catch (e) {
            console.log(e);
        }

        return [false, "데이터를 조회할 수 없습니다.", null];
    }

    static async requestCancelVacations(requestIDs) {
        try {
            const jsonData = JsonManager.makeRequestCancelVacations(requestIDs);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return [result.success, result.message, result.history, result.historyNextYear];
            }
        }
        catch (e) {
            console.log(e);
        }

        return [false, "휴가취소가 실패하였습니다.", null, null];
    }

    static async requestHolidays(year, month = null) {
        try {
            const jsonData = JsonManager.makeRequestHolidays(year, month);

            const res = await fetch('/api/Vacation', {
                method: 'post',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: jsonData
            });

            if (res.ok) {
                const result = await res.json();
                return [result.success, result.message, result.holidays];
            }
        }
        catch (e) {
            console.log(e);
        }

        return [false, "공휴일 정보를 얻어오는데 실패하였습니다.", null];
    }
}